Assets/Scripts/CharacterMechanics/CharacterMovement.cs

```diff
@@ -0,0 +1,584 @@
+    float JumpPower
+    {
+        get
+        {
+            float height = JumpHeight + (Jumps - ExtraJumpsRemaining - 1) * JumpHeightBonus;
+            return Mathf.Sqrt(2 * CharacterGravity * height);
+        }
+    }
+    CharacterController Controller => GetComponent<CharacterController>();
+    Vector3 ControllerWorldPosition => transform.position + Controller.center;
+
+    public Vector3 TopSphereCenter =>
+        ControllerWorldPosition + GravityUpDirection * (Controller.height / 2 - Controller.radius);
+    public Vector3 BottomSphereCenter =>
+        ControllerWorldPosition - GravityUpDirection * (Controller.height / 2 - Controller.radius);
+
+    public bool IsFalling => VerticalSpeed < 0;
+    public bool IsRising => VerticalSpeed > 0;
+    public bool LateralMovementEnabled { get; set; } = true;
+    public bool RotationEnabled { get; set; } = true;
+    public bool VerticalMovementEnabled
+    {
+        get { return GravityEnabled && StateChangeEnabled; }
+        set
+        {
+            GravityEnabled = value;
+            StateChangeEnabled = value;
+
+            VerticalState = value ? VerticalMovementState.Falling : VerticalMovementState.Grounded;
+            VerticalSpeed = 0;
+        }
+    }
+    public bool GravityEnabled { get; set; } = true;
+    public bool StateChangeEnabled { get; set; } = true;
+    #endregion
+
+    #region Private Properties
+    float CharacterGravity => Physics.gravity.magnitude * GravityMultiplier;
+    float TimeSinceGrounded => IsOnStableGround() ? 0 : Time.time - LastTimeGrounded;
+    bool IsHittingHead =>
+        Mathf.Abs(Vector3.Dot(Controller.velocity, GravityUpDirection)) < dx && IsRising;
+    #endregion
+
+    #region Constants
+    const float dx = 0.01f;
+    #endregion
+
+    [NonSerialized]
+    public Vector3 GravityUpDirection = new(0, 1, 0);
+
+    //[NonSerialized]
+    public float VerticalSpeed = 0;
+
+    public Vector3 RawMovementDirection = new();
+    public Vector3 RawFacingDirection = new();
+    public Func<Vector3, float, Vector3> MovementDirectionMiddleware = (v, dt) => v;
+    public Func<Vector3, float, Vector3> FacingDirectionMiddleware = (v, dt) => v;
+    public Vector3 FacingDirection = new();
+    public Vector3 MovementDirection = new();
+
+    [NonSerialized]
+    public int ExtraJumpsRemaining = 0;
+
+    [NonSerialized]
+    public Vector3 AdditionalImpulse = Vector3.zero;
+
+    //[NonSerialized]
+    public VerticalMovementState VerticalState = VerticalMovementState.Falling;
+
+    float LastTimeGrounded = 0;
+
+    #region Platform Tracking
+    // the last platform the character landed on
+    GameObject LastPlatform;
+
+    // the position of the last platform on the previous frame
+    Vector3 LastPlatformLastPosition;
+
+    // the velocity of the last platform on the previous frame
+    Vector3 LastPlatformLastVelocity;
+    float LastPlatformLastTime;
+    #endregion
+
+    readonly Maid maid = new();
```

Comment:
The Maid is really, really nice - I know it wasn't strictly necessary but I think that adding it makes the code much simpler and easier to deal with